capstone/src/test/java/com/google/step/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,125 @@
+package com.google.step.servlets;
+
+import static org.junit.Assert.*;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.repackaged.com.google.gson.JsonObject;
+import com.google.appengine.repackaged.com.google.gson.JsonParser;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.meterware.httpunit.GetMethodWebRequest;
+import com.meterware.httpunit.WebRequest;
+import com.meterware.httpunit.WebResponse;
+import com.meterware.servletunit.ServletRunner;
+import com.meterware.servletunit.ServletUnitClient;
+import java.io.IOException;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.xml.sax.SAXException;
+
+public class LoginServletTest {
+    private final LocalServiceTestHelper helper =
+            new LocalServiceTestHelper();
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    @Test
+    public void testIsLoggedIn() {
+        // Check if this works to set the user as logged in and not the admin
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(true);
+        UserService userService = UserServiceFactory.getUserService();
+        assertTrue(userService.isUserLoggedIn());
+        assertFalse(userService.isUserAdmin());
+    }
+
+    @Test
+    public void testIsNotLoggedIn() {
+        // Check if it works to show the user is not logged in
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(false);
+        UserService userService = UserServiceFactory.getUserService();
+        assertFalse(userService.isUserLoggedIn());
+        // Can't also check if the user is an admin because the method will throw an error since the
+        // user is not even logged in.
+    }
+
+    @Test
+    public void testDoGetWithoutLogIn() throws IOException, SAXException {
+        // Testing doGet method without having the user logged in.
+        // Result: {"LoggedIn":"false","loginURL":"someURL"}
+
+        ServletRunner sr = new ServletRunner();
+        // The first parameter has to be the same as whatever goes after the '/' in the urlString for
+        // the GetMethodWebRequest so for example if the first parameter is 'restaurant' then the link
+        // would be: 'http://localhost:8080/restaurant' even though the actual servlet may be called
+        // something else.
+        sr.registerServlet("login", LoginServlet.class.getName());
+        ServletUnitClient sc = sr.newClient();
+
+        // This can throw the SAXException.
+        WebRequest request = new GetMethodWebRequest("http://localhost:8080/login");
```

Comment:
Does this framework not offer a way to run a servlet with a random port (whatever's available at the time)?

What if you're just running the test suite for your entire system, and many servlet unit tests want to run in parallel? Can they all bind to the same port at the same time?

Reply:
Yes they should be all able to bind to the same port to my understanding. But I can look to migrating into mockito for the future.